bpf-socket-bind: fix unexpected behavior with either 0 allow or deny rules

networkException opened the pull request:
This pull request fixes an issue where, when not specifiying either at least one `SocketBindAllow` or `SocketBindDeny` rule, behavior for the bind syscall filtering would be unexpected.

For example, when trying to bind to a port with only "SocketBindDeny=any" given, the syscall would succeed:

```bash
systemd-run -t -p "SocketBindDeny=any" nc -l 8080
```

Expected with this set of rules (also in accordance with the documentation) would be an Operation not permitted error.

This behavior occurs because a default initialized socket_bind_rule struct matches what "any" represents. When creating the bpf list all elements get default initialized, as such represeting "any". Seemingly it is necressarry to set the size of the map to at least one, as such if no allow rule is given default initialization and minimal map size cause one any allow rule to be in the map, causing the behavior observed above.

This pull request solves this by adding an "initialized" boolean to the struct and only setting it to true when a rule was actually given.

Resolves #30556

bluca commented:
please add test coverage in src/test/test-socket-bind.c or test/units/testsuite-07.exec-context.sh

poettering commented on src/core/bpf/socket_bind/socket-bind-api.bpf.h:
```diff
@@ -46,6 +46,7 @@ struct socket_bind_rule {
         __u32 protocol;
         __u16 nr_ports;
         __u16 port_min;
+        _Bool initialized;
```
any particular reason why you use `_Bool` rather than `bool` here?

poettering replied:
hmm, so changing the size of the structure sucks for two reasons: we cannot change the structure for running systems really, i.e. "systemctl daemon-reexec" just breaks here. And there might be a ton of entries in some setups, hence increasing it in size should maybe not be done willy-nilly...

wouldn't it be a better approach to just always insert at least one entry? i.e. adding the right one if none is configured?

networkException replied:
> any particular reason why you use `_Bool` rather than `bool` here?

I had some wrong assumptions about the other types in the struct. I'm not too familiar with C and sometimes it shows when experience with other languages isn't enough :sweat_smile:

> hmm, so changing the size of the structure sucks for two reasons: we cannot change the structure for running systems really, i.e. "systemctl daemon-reexec" just breaks here. And there might be a ton of entries in some setups, hence increasing it in size should maybe not be done willy-nilly...

I see

> wouldn't it be a better approach to just always insert at least one entry? i.e. adding the right one if none is configured?

I initially tried just inserting null values which didn't work (correct me if there's a way, that would make things a lot easier)

As far as I can see there is (with the current matching) no way to express "match no address".

As the matching code currently ignores anything but the 8 least most significant bits from the address family passed to the syscall (and anything not `AF_INET` or `AF_INET6` for either `user_family` or `family`) anyways we could put a constant in the 32 bit unsigned rule indicating this.

networkException commented:
I removed the additional boolean field as per https://github.com/systemd/systemd/pull/31707#discussion_r1519879219 and introduced a magic value to be interpreted as "match nothing" instead. The magic is set into the `address_family` field if no rule is given for either allow or deny, replacing the default initialized value.

yuwata commented on src/core/bpf-socket-bind.c:
```diff
@@ -32,6 +32,18 @@ static int update_rules_map(
         assert(map_fd >= 0);
 
+        if (!head) {
+                struct socket_bind_rule val = {
```
Maybe better to add `const` and `static` attributes?

yuwata reviewed (approved):
LGTM.

keszybz commented on src/core/bpf/socket_bind/socket-bind-api.bpf.h:
```diff
@@ -49,3 +52,4 @@ struct socket_bind_rule {
 };
 
 #define SOCKET_BIND_MAX_RULES 128
+#define SOCKET_BIND_RULE_AF_MATCH_NOTHING 4294967295
```
Please write this as `(UINT32_MAX - 1)` instead of a number.

yuwata replied:
UINT32_MAX ?

networkException replied:
Yes `4294967295` would be `UINT32_MAX`. I've updated the definition

keszybz commented on src/core/bpf-socket-bind.c:
```diff
@@ -32,6 +32,18 @@ static int update_rules_map(
+        if (!head) {
+                static const struct socket_bind_rule val = {
+                        .address_family = SOCKET_BIND_RULE_AF_MATCH_NOTHING,
+                        .protocol = 0,
+                        .nr_ports = 0,
+                        .port_min = 0,
```
Please drop those three lines, the defaults are 0.

keszybz reviewed (changes requested):
LGTM. Minor style tweaks.

keszybz commented on src/core/bpf-socket-bind.c:
```diff
@@ -32,6 +32,15 @@ static int update_rules_map(
 
+        if (!head) {
+                static const struct socket_bind_rule val = {
+                        .address_family = SOCKET_BIND_RULE_AF_MATCH_NOTHING
```
Hmm, we lost the comma here…
